Allow up to 10 speeds in UnitarySystemPerformance:Multispeed

rraustad opened the pull request:
Pull request overview
---------------------
The UnitarySystemPerformance:Multispeed object allows up to 4 heating and cooling speed ratios. The variable speed DX cooling coil allows up to 10 speeds. The IDFEditor cannot open an input file with a 10-speed DX cooling coil that uses a multispeed performance object unless the IDD is extended to include 10 fields. For this reason, the number of fields in the UnitarySystemPerformance:Multispeed object have been increased to 10.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus

NFP Check list:
- [x] Extend number of speeds in IDD to 10. See suggested IDD in draft NFP.
- [x] Test that number of speeds < 10 works with modified IDD
- [x] Test input file without UnitarySystemPerformance:Multispeed object (should already work).
- [x] Test use of UnitarySystemPerformance:Multispeed object with 10-spd DX coil object
- [x] Update ExpandObjects.f90
- [x] Update Version Translator
- [x] Update Object Transition Rules
- [x] Add unit test
- [x] Update IO Ref

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Code style (parentheses padding, variable names)
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check -- verify this
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add rules to spreadsheet
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Documentation changes in place
 - [ ] Changed docs build successfully
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

rraustad committed:
NFP Initial draft

rraustad committed:
Add Final NFP and Design Doc

rraustad commented:
@mjwitte shouldn't all input fields with \units C have a companion \ip-units F ?

    AirLoopHVAC:UnitarySystem,
      N2 , \field Minimum Supply Air Temperature
        \type real
        \units C
        \minimum 0.0
        \maximum 20.0
        \default 2.0
        \autosizable
        \note When Use DOAS DX Cooling Coil is specified as Yes, Minimum Supply Air Temperature

mjwitte commented:
No, there's a set of default conversions near the top of the IDD.
```
! Default IP conversions (no \ip-units necessary)
!     m                      =>   ft                  3.281
!     W                      =>   Btu/h               3.412
!     m3/s                   =>   ft3/min             2118.6438
!     C                      =>   F                   1.8 (plus 32)
. . . 
```

rraustad commented:
So these should be removed? There's only about 50 instances.

WindowMaterial:GlazingGroup:Thermochromic,
  N1 , \field Optical Data Temperature 1
       \begin-extensible
       \required-field
       \units C
       \ip-units F

EvaporativeFluidCooler:SingleSpeed,
  N9 , \field Design Entering Water Temperature
       \type real
       \units C
       \minimum> 0.0
       \ip-units F

mjwitte commented:
It doesn't hurt anything.  I wouldn't bother.

rraustad committed:
Update IDD and IO Ref

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into New-Feature---Increase-speeds-in-multispeed-object

rraustad committed:
Add no load air flow rate ratio to MS performance object

rraustad committed:
Transition Rules 8-7-0 to 8-8-0. ExpandObjects update.

rraustad committed:
Update speed 5-10 in epfilter.f90

rraustad committed:
Update unit tests

mjwitte commented on doc/input-output-reference/src/overview/group-unitary-equipment.tex:
```diff
@@ -718,27 +723,75 @@ \subsubsection{Inputs}\label{inputs-1-046}
 \paragraph{Field: Heating Speed 2 Supply Air Flow Ratio}\label{field-heating-speed-2-supply-air-flow-ratio}
 
-This numeric field defines the ratio of supply air flow rate leaving the unitary system to the maximum air flow rate specified in the coil object at maximum speed when the heating coil and/or supplemental heater are operating at Speed 2. Values must be greater than 0. The entered value must be greater or equal to the flow rate ratio specified for heating speed 1.
+This numeric field defines the ratio of supply air flow rate leaving the unitary system to the maximum air flow rate specified in the coil object at maximum speed when the heating coil and/or supplemental heater are operating at Speed 2. Values must be greater than 0. The entered value must be greater or equal to the flow rate ratio specified for heating speed 1. If the `Number of Speeds for Heat ...
```
Would prefer to collapse all the speeds into one description labeled ". . . Speed < x > . . . " similar to Coil:Cooling:DX:MultiSpeed.

rraustad replied:
Done

![msheatcoolspeedx](https://cloud.githubusercontent.com/assets/5778156/26367216/d4a42d84-3fbb-11e7-826e-e95a98435d9b.png)

rraustad replied:
Where did this go? I obviously changed the doc as suggested, but it's not showing up here. Did I not commit this change? Maybe did locally and didn't push.

rraustad replied:
I think I recall that I committed this at work but didn't push since I was waiting for other comments.

rraustad commented:
This PR still needs a unit test and a little more testing.

rraustad committed:
Update IO Ref to generalize speed X for 10 speed coils

nrel-bot-3 commented:
@rraustad @lgentile it has been 16 days since this pull request was last updated.

nrel-bot-2 commented:
@rraustad @lgentile it has been 15 days since this pull request was last updated.

nrel-bot-3 commented:
@rraustad @lgentile it has been 14 days since this pull request was last updated.

rraustad commented:
Waiting on #6107 before creating a unit test and completing this issue.

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into New-Feature---Increase-speeds-in-multispeed-object

Conflicts:
	src/Transition/InputRulesFiles/Rules8-7-0-to-8-8-0.md

rraustad committed:
Units tests

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into New-Feature---Increase-speeds-in-multispeed-object

rraustad commented:
I'm seeing more of these. I assume it's the CI. Is this something that can be corrected?

unknown file: error: C++ exception with description "invalid string position" thrown in the test bod

rraustad commented:
OK, so I went ahead and added in the UnitarySystemPerformance:Multispeed object to the 10-spd VS coil (because it's useless to increase speeds to 10 and not be able to use them). This is working as I expect. Someone remind me to change docs again if I need to.

rraustad committed:
Add MSHP design spec object to VS DX coils

rraustad commented on tst/EnergyPlus/unit/HVACUnitarySystem.unit.cc:
```diff
@@ -723,50 +723,52 @@ TEST_F( ZoneUnitarySystemTest, UnitarySystem_MultispeedPerformance ) {
-	// Coil:*:DX:VariableSpeed coils DO NOT use multispeed performance object information
-	// when and if this is corrected, these numbers will change based on the inputs for the MS performance object
-	EXPECT_NEAR( UnitarySystem( 1 ).CoolVolumeFlowRate( 1 ), 0.638152, 0.000001 );
+	EXPECT_NEAR( UnitarySystem( 1 ).CoolVolumeFlowRate( 1 ), 0.177264, 0.000001 );
```
See how much these changed. Before, the coil object determined what the flow rate would be, Now the MSHP design spec object sets the flow and I set it up as 10% per speed (1.772 * 10% = 0.1772).

mjwitte replied:
@rraustad Remind me what the difference is between these two approaches to controlling the VS coil.  Briefly, if that's possible.

rraustad commented:
The VS coil is a rouge. It requires hard-sized speed data, then calculates an intermediate TotalCapacity based on total zone flow rate. Then takes the hard-sized speed data, the airflow to capacity ratio, and uses that data to back calculate air flow rates. Then takes another step to use those air flow rates to calculate again the TotalCapacity. This makes the final result exactly scaled to the original hard-sized speed data, but it also means I can't use the UnitarySystemPerformance:Multispeed to adjust air flow rates.

Example:

    "  Coil:Cooling:DX:VariableSpeed,",

    "    autosize,                !- Rated Total Cooling Capacity At Selected Nominal Speed Level {w}",
    "    autosize,                !- Rated Volumetric Air Flow Rate At Selected Nominal Speed Level {m3/s}",

    "    33861.72,                !- Speed 1 Reference Unit Total Cooling Capacity At Rated Conditions {w}",
    "    0.70,                    !- Speed 1 Reference Unit Sensible Heat Ratio At Rated Conditions {dimensionless}",
    "    4.328973,                !- Speed 1 Reference Unit COP At Rated Conditions {dimensionless}",
    "    1.396964,                !- Speed 1 Reference Unit Air Flow Rate At Rated Conditions {m3/s}",

With autosized Total cooling capacity = 100,000 W, then air flow rate at speed 1 must be equal to 100,000 * ( 1.396964 / 33861.72 )

So I thought that scaling the air flow rate using the UnitarySystemPerformance:Multispeed object and then calculating the capacity at each speed using TotalCapacity = AdjustedAirFlowRate * (33861.72 / 1.396964 ) would be a good compromise. It allows the user to model the DX VS coil exactly as originally intended, and if they include the UnitarySystemPerformance:Multispeed object, to scale coil sizing a little differently than the original model.

So the end result is a little more flexibility in sizing the coil performance.

rraustad committed:
Autosize UnitarySystemPerformance:Multispeed air flow ratios before calling VS DX coils

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into New-Feature---Increase-speeds-in-multispeed-object

# Conflicts:
#	src/Transition/InputRulesFiles/Rules8-7-0-to-8-8-0.md

rraustad committed:
Merge branch 'New-Feature---Increase-speeds-in-multispeed-object' of https://github.com/NREL/EnergyPlus into New-Feature---Increase-speeds-in-multispeed-object

rraustad committed:
Fix for failed VS DX coil example files

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into New-Feature---Increase-speeds-in-multispeed-object

mjwitte commented:
@rraustad Please edit the title and description for this PR.  Also, this branch has a small conflict now in the rules file.

rraustad commented:
Work for tonight at home:
- [x] Change PR title
- [x] fix conflict
- [x] Change unit test line causing diff's to EXPECT_NEAR
- [x] c;lean up indentation in unit test

rraustad committed:
Change unit test to use EXPECT_NEAR and clean up indentation in unit test.
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into New-Feature---Increase-speeds-in-multispeed-object

Conflicts:
	src/Transition/InputRulesFiles/Rules8-7-0-to-8-8-0.md

rraustad commented:
I am tired of this indentation issue with the unit tests. Some use spaces, some use tabs. If you copy paste from one to another and then don't notice, the diff's show as indentation issues when in an editor they all line up. Doesn't matter what I do, change all to tabs, or change all to spaces, someone will disagree. Someone needs to write a script to fix this so I stop wasting time trying to fix what doesn't need fixing.

Myoldmopar commented on src/EnergyPlus/HVACUnitarySystem.cc:
```diff
@@ -3023,38 +3052,39 @@ namespace HVACUnitarySystem {
 			MaxSpeedNum = 0; // set to 0 so we know if the inputs are valid
 
-			if ( mod( double( NumNumbers - 2 ), 2.0 ) > 0.0 ) {
+			if ( mod( double( NumNumbers - 3 ), 2.0 ) > 0.0 ) {
```
Interesting.  I know you didn't add it here, so no big deal, but I'm not used to seeing `mod` as a function. Usually an operator is used:

```
(NumNumbers-3) % 2 > 0
```

Nothing to worry about here, just a new one for me.

Myoldmopar commented on idd/Energy+.idd.in:
```diff
@@ -54474,47 +54481,143 @@ UnitarySystemPerformance:Multispeed,
        \note operation or specify autosize. This value is the ratio of air flow
        \note at this speed to the maximum air flow rate.
-  N10; \field Cooling Speed 4 Supply Air Flow Ratio
+  N23; \field Cooling Speed 10 Supply Air Flow Ratio
```
Everything looks good here :+1:

Myoldmopar commented on src/EnergyPlus/HVACUnitarySystem.cc:
```diff
@@ -1794,7 +1795,7 @@ namespace HVACUnitarySystem {
 		// SUBROUTINE LOCAL VARIABLE DECLARATIONS:
 		int Iter; // iteration count
-		Real64 MulSpeedFlowScale; // variable speed air flow scaling factor
+//		Real64 MulSpeedFlowScale; // variable speed air flow scaling factor
```
Remove if any other code changes are needed.

rraustad replied:
Shoot, missed this.

Myoldmopar commented on src/EnergyPlus/HVACUnitarySystem.cc:
```diff
@@ -1964,6 +1965,10 @@ namespace HVACUnitarySystem {
+				if ( UnitarySystem( UnitarySysNum ).CoolingCoilType_Num == Coil_CoolingAirToAirVariableSpeed ) {
+					DataTotCapCurveIndex = GetVSCoilCapFTCurveIndex( UnitarySystem( UnitarySysNum ).CoolingCoilIndex, ErrFound );
+					DataIsDXCoil = true;
+				}
```
This seems unrelated.  Was this not needed before?  Or is this added here as a fix for something else.

rraustad replied:
The VS DX coil never used the TotCapFTemp curve to adjust coil capacity during sizing. This should have been here all along.

Myoldmopar commented on src/EnergyPlus/HVACUnitarySystem.cc:
```diff
@@ -2216,13 +2227,22 @@ namespace HVACUnitarySystem {
+
 			for ( Iter = 1; Iter <= UnitarySystem( UnitarySysNum ).NumOfSpeedCooling; ++Iter ) {
-				UnitarySystem( UnitarySysNum ).CoolVolumeFlowRate( Iter ) = VarSpeedCoil( UnitarySystem( UnitarySysNum ).CoolingCoilIndex ).MSRatedAirVolFlowRate( Iter ) * MulSpeedFlowScale;
+				UnitarySystem( UnitarySysNum ).CoolVolumeFlowRate( Iter ) = VarSpeedCoil( UnitarySystem( UnitarySysNum ).CoolingCoilIndex ).MSRatedAirVolFlowRate( Iter );
```
Could you explain why this multiplier isn't needed anymore?  Mostly just me wanting to understand this section of code better.

rraustad replied:
@nealkruis worked on this issue a while back. See PR #5723 line 3000 of PackagedTerminalHeatPump. I am trying to gently pull back on the VS DX coil so it acts more like other coil models when it comes to sizing. The VS DX coil model back calculates air flow based on capacity, then over in the parent there also needs to be back flips to get to the data needed. This is very confusing and needs to be corrected. As for why this was removed here, the parent object will now use the coil air flow rate instead of back flipping to something other that what the coil uses.

Myoldmopar commented on src/Transition/InputRulesFiles/Rules8-7-0-to-8-8-0.md:
```diff
@@ -8,17 +8,22 @@ This was previously an Excel workbook that made for very difficult version contr
+
+Fields 1-4 remain the same.  
+After F4, insert one new blank field that represents the No Load Supply Air Flow Rate Ratio. 
+Shift all later fields down by one.
```
I'll work up this transition rule, probably tomorrow morning.

Myoldmopar replied:
OK, there is something sketchy going on.  I went to add this transition rule to the Fortran source, and found that it wouldn't compile, because there was already a rule in there for this object.  I thought, OK, this is probably just one of those uncommon situations where a second transition rule is applied, and we need to carefully merge them together.  But then I noticed they were the exact same rule.  Weird.  I went back and found I had added the transition rule in as part of the many transition rules for #6158 [here](https://github.com/NREL/EnergyPlus/pull/6158/files#diff-1a3d976fbe90f0b4b2131917bff01b5fR369).  I added it there because it was one of the requested additions in the Rules file [there](https://github.com/NREL/EnergyPlus/pull/6158/files#diff-84d66175020363cd70e1e9161a145675R23).  But I can't figure out why that was in that pull request, when it's really a part of this PR.

Either way, the "Rules" file and the Fortran transition source both have the no-load field in it.  And now with this PR, the IDD and code finally have the changes in them.  It seems like everything will be back to compatible once this PR goes in.  Can you confirm?

Myoldmopar replied:
I went ahead and pushed up the rules file after resolving the merge conflict.  If you help me understand what happened with the transition rule in the other PR, and it looks like that is good to go, then this can go in.

rraustad replied:
I have had to merge the rules file several times. I can't say that I paid close attention to what was being merged as how it needed to be merged. I am a little surprised this snuck in. It might be because I saved a branch and then popped it back in the wrong branch. I've done this several times lately and have been switching branches a lot to try to keep up. That's my only guess.

Myoldmopar replied:
Thanks for the reply.  I built it locally and I am getting a failure in the new unit test.  Says it failed to process the input.  I'm investigating.  If you could peek at the inputs in that unit test as well that would be great.  My first guess was there was a copy/paste and the new no-load field you added here wasn't in there or something, but it looks like it is there, so I'm not sure yet.

rraustad replied:
I'll run here, you can move on if you have other things to do.

Myoldmopar reviewed (changes requested):
This still needs transition rules, so I'm "requesting changes" ... mostly to myself :upside_down_face: 

There was one commented line I'll remove, but otherwise it looks good.

I asked for clarification on a couple of the code changes, so if you can just describe those to help me understand, that would be great and we can get this in quickly.

Myoldmopar committed:
Merge develop; resolve conflict

rraustad commented:
Unit test is failing because the VS DX coil don't have the minimum OAT for compressor operation field inserted in the object. I'll fix.

rraustad committed:
Update new unit test to include Min OAT for Compressor operation input field. Other clean up in unit test.

rraustad commented on tst/EnergyPlus/unit/HVACUnitarySystem.unit.cc:
```diff
@@ -228,6 +245,538 @@ class ZoneUnitarySystemTest: public EnergyPlusFixture {
+	// variable speed coils size air flow differently than other models. The design air volume flow rate is back calculated from design capacity
+	EXPECT_EQ( VarSpeedCoil( 1 ).RatedAirVolFlowRate, VarSpeedCoil( 1 ).RatedCapCoolTotal * VarSpeedCoil( 1 ).MSRatedAirVolFlowPerRatedTotCap( 10 ) );
+	EXPECT_NEAR( VarSpeedCoil( 1 ).RatedAirVolFlowRate, 1.77264, 0.00001 ); // different than unitary system air volume flow rate
+	EXPECT_NEAR( VarSpeedCoil( 2 ).RatedAirVolFlowRate, 1.70, 0.01 ); // VS DX heating coil was not autosized
```
These tests highlight how the VS DX coil model air flow sizing is so different from other models. The autosized air flow rate for unitary system = 1.5, the autosized air flow rate for VS DX cooling = 1.772. Not sure we can get to something that resolves the difference because of how the VS DX coil model uses hard-sized inputs for air flow and capacity at each speed to "scale" the autosized data. But we can try to make this better. It's important because Trane is relying on the VS DX coil model for it's GUI.

rraustad commented:
I can't find this unit test: MultiSpeedHeatPump_MultiSolvers ?? I've pulled both this branch and develop.

Myoldmopar commented:
It's not a unit test.  It's an example file.  I see the failure.  It's likely just needing an input added, I'll fix it up.

Myoldmopar committed:
Fix MultiSpeed example file input

Myoldmopar commented:
Done.  It runs now.  Since it was the only failure, I'm merging now.  Thanks @rraustad
